### PR TITLE
Add secret support to Terraform Providers (backend updates)

### LINKED ISSUE #6539: Allow any Terraform provider to be configured and used in a Recipe
## Overview of feature request

Today we support Terraform templates for Recipes. For the Azure, Kubernetes, and AWS providers we automatically setup the provider configuration with the identity of where to deploy and we pass in where to deploy via the `context` parameter.

Additionally, any additional provider can be used in a Recipe, however there is no way to securely configure the provider with credentials. For example, there is no way to use the GCP Terraform Provider in a Recipe without hard-coding the credentials in the Recipe module.

This feature is for allowing credentials to be securely configured and passed into the Recipe so any Terraform Provider can be used.

## Acceptance criteria

Once this feature is implemented I should be able to use the GCP, Oracle Cloud, or any other Terraform provider and pass in the credentials. I should be able to follow documentation that walks me through the steps I should take to accomplish this, with snippets showing an example.

There should be unit tests confirming that credentials are passed in correctly.

## Additional context

There are a couple ways to implement this, as each Terraform provider handles auth in various ways (_managed identities, service principals, IAM roles, etc._). We should take a look at how most providers handle credentials and see what makes sense to implement to cast the widest nest. Could be parameters that are passed in with the credentials, or could be a Terraform RC file pass through.

## Community ask

Please comment below with any Terraform modules that you would like to use.

[AB#9883](https://dev.azure.com/azure-octo/e61041b4-555f-47ae-95b2-4f8ab480ea57/_workitems/edit/9883)

### PR DESCRIPTION
**Pull request by lakshmimsft**
# Description

Added backend updates for secret support to Terraform Providers configuration. This includes processing secrets in recipeConfig under specific Provider configurations and environment variables. It also updates existing functions which add secrets to .gitConfig for private terraform modules.

Design Document: https://github.com/radius-project/design-notes/pull/47

## Type of change

- This pull request fixes a bug in Radius and has an approved issue #6539 .

Fixes: #6539

### CODE REVIEW

File: pkg/recipes/terraform/config/providers/types.go [hunk 1]
```diff
@@ -47,26 +49,57 @@ func GetUCPConfiguredTerraformProviders(ucpConn sdk.Connection, secretProvider *
...
+
+// extractSecretsFromConfig extracts secrets for env recipe configuration from the secrets data input and updates the currentConfig map.
+func extractSecretsFromConfig(currentConfig map[string]any, recipeConfigSecrets map[string]datamodel.SecretReference, secrets map[string]map[string]string) error {
+	// Extract secrets from configDetails if they are present
+	for secretName, secretReference := range recipeConfigSecrets {
+
+		// Extract secret value from the secrets data input
+		if secretSource, ok := secrets[secretReference.Source]; ok {
+			if secretValue, ok := secretSource[secretReference.Key]; ok {
+				currentConfig[secretName] = secretValue
```
Comment (ytimocin) [inline pkg/recipes/terraform/config/providers/types.go:95]: Can this be overridden by any chance? I mean, can `currentConfig[secretName]` be already not empty at this point?

File: pkg/recipes/terraform/execute.go [hunk 2]
```diff
@@ -91,7 +91,7 @@ func (e *executor) Deploy(ctx context.Context, options Options) (*tfjson.State,
 
 	if options.EnvConfig != nil {
 		// Set environment variables for the Terraform process.
-		err = e.setEnvironmentVariables(tf, &options.EnvConfig.RecipeConfig)
+		err = e.setEnvironmentVariables(tf, &options.EnvConfig.RecipeConfig, options.Secrets)
```
Comment (ytimocin) [inline pkg/recipes/terraform/execute.go:94]: Why not send the whole `options`?

File: pkg/recipes/terraform/types.go [hunk 3]
```diff
@@ -60,6 +60,11 @@ type Options struct {
 
 	// ResourceRecipe is recipe metadata associated with the Radius resource deploying the Terraform recipe.
 	ResourceRecipe *recipes.ResourceMetadata
+
+	// Secrets represents a map of secrets required for recipe execution.
+	// The outer map's key represents the secretStoreIDs while
+	// while the inner map's key-value pairs represent the [secretKey]secretValue.
+	Secrets map[string]map[string]string
```
Comment (ytimocin) [inline pkg/recipes/terraform/types.go:67]: Can we have an example in the doc?

File: pkg/recipes/driver/gitconfig.go [hunk 4]
```diff
@@ -149,3 +144,20 @@ func GetGitURL(templatePath string) (*url.URL, error) {
 
 	return url, nil
 }
+
+// wrapper function to add secrets to .gitconfig if applicable.
+func addSecretsToGitConfigIfApplicable(secretStoreID string, secretData map[string]map[string]string, requestDirPath string, templatePath string) error {
+	if secretStoreID != "" && secretData != nil {
```
[the remark this answers is not in the document]
Comment (lakshmimsft) [reply pkg/recipes/driver/gitconfig.go]: no, it was supposed to be &&, we return nil if this condition is not met, as u mention.  
I refactored the function for better readability.

File: pkg/recipes/engine/engine.go [hunk 5]
```diff
@@ -226,24 +226,30 @@ func (e *engine) getDriver(ctx context.Context, recipeMetadata recipes.ResourceM
 	return definition, driver, nil
 }
 
-func (e *engine) getRecipeConfigSecrets(ctx context.Context, driver recipedriver.Driver, configuration *recipes.Configuration, definition *recipes.EnvironmentDefinition) (v20231001preview.SecretStoresClientListSecretsResponse, error) {
-	secrets := v20231001preview.SecretStoresClientListSecretsResponse{}
+func (e *engine) getRecipeConfigSecrets(ctx context.Context, driver recipedriver.Driver, configuration *recipes.Configuration, definition *recipes.EnvironmentDefinition) (secretData map[string]map[string]string, err error) {
 	driverWithSecrets, ok := driver.(recipedriver.DriverWithSecrets)
 	if !ok {
-		return secrets, nil
+		return nil, nil
```
[the remark this answers is not in the document]
Comment (lakshmimsft) [reply pkg/recipes/engine/engine.go:232]: this is existing code. Here the driver does not implement DriverWithSecrets interface and we explicitly return nil. I think it is clearer. We could also return secretData which is not yet initialized and it would amount to returning nil but for better clarity sticking with nil.

File: pkg/recipes/terraform/config/config.go [hunk 6]
```diff
@@ -230,9 +230,12 @@ func newModuleConfig(moduleSource string, moduleVersion string, params ...Recipe
 // The function returns a map where the keys are provider names and the values are slices of maps.
 // Each map in the slice represents a specific configuration for the corresponding provider.
 // This structure allows for multiple configurations per provider.
-func getProviderConfigs(ctx context.Context, requiredProviders map[string]*RequiredProviderInfo, ucpConfiguredProviders map[string]providers.Provider, envConfig *recipes.Configuration) (map[string][]map[string]any, error) {
+func getProviderConfigs(ctx context.Context, requiredProviders map[string]*RequiredProviderInfo, ucpConfiguredProviders map[string]providers.Provider, envConfig *recipes.Configuration, secrets map[string]map[string]string) (map[string][]map[string]any, error) {
```
[the remark this answers is not in the document]
Comment (lakshmimsft) [reply pkg/recipes/terraform/config/config.go:233]: staying with current inputs. will revisit if necessary.

File: pkg/recipes/terraform/config/config.go [hunk 7]
```diff
@@ -297,3 +300,31 @@ func (cfg *TerraformConfig) AddOutputs(localModuleName string) error {
...
+}
+
+// updateSecretIDs is helper function to update the inputSecretIDs map
+func updateSecretIDs(inputSecretIDs map[string][]string, secretSourceId, key string) {
+	if _, ok := inputSecretIDs[secretSourceId]; !ok {
+		inputSecretIDs[secretSourceId] = []string{key}
+	} else {
+		inputSecretIDs[secretSourceId] = append(inputSecretIDs[secretSourceId], key)
+	}
+}
```
[the remark this answers is not in the document]
Comment (lakshmimsft) [reply pkg/recipes/terraform/config/config.go]: updated to addSecretKeys. open to suggestions.

File: pkg/recipes/terraform/config/providers/types.go (hunk 1 above)
Comment (lakshmimsft) [reply pkg/recipes/terraform/config/providers/types.go:94]: This is a good question. The one way that it is possible is if the user puts in the same key in provider configuration and under secrets for a provider configuration. eg bicep input:
```
          azurerm: [
            {
              subscriptionid: 1234
              tenant_id: '745fg88bf-86f1-41af'
              client_id: 'abc123'
              secrets: { 
                client_id: {
                  source: secretStoreConfig.id
                  key: 'secretkey.client_id'
                }
                password: {
                  source: secretStoreConfig.id
                  key: 'secretkey.password'
                }
              }
            },
```
 The current code will have the Secret value override any existing values for the same key in the provider configuration. We can review if this is desired/acceptable behaviour.

File: pkg/recipes/driver/gitconfig.go [hunk 8]
```diff
@@ -71,8 +66,8 @@ func getURLConfigKeyValue(secrets v20231001preview.SecretStoresClientListSecrets
 // Retrieves the git credentials from the provided secrets object
 // and adds them to the Git config by running
 // git config --file .git/config url<template_path_domain_with_credentails>.insteadOf <template_path_domain>.
-func addSecretsToGitConfig(workingDirectory string, secrets v20231001preview.SecretStoresClientListSecretsResponse, templatePath string) error {
-	if !strings.HasPrefix(templatePath, "git::") || reflect.DeepEqual(secrets, v20231001preview.SecretStoresClientListSecretsResponse{}) {
+func addSecretsToGitConfig(workingDirectory string, secrets map[string]string, templatePath string) error {
+	if !strings.HasPrefix(templatePath, "git::") || secrets == nil || len(secrets) == 0 {
```
[the remark this answers is not in the document]
Comment (lakshmimsft) [reply pkg/recipes/driver/gitconfig.go:70]: discussed with yetkin. current code is good for now.

File: pkg/recipes/engine/engine.go [hunk 9]
```diff
@@ -226,24 +226,30 @@ func (e *engine) getDriver(ctx context.Context, recipeMetadata recipes.ResourceM
...
+		secretData, err = e.options.SecretsLoader.LoadSecrets(ctx, secretStoreIDResourceKeys)
 		if err != nil {
-			return v20231001preview.SecretStoresClientListSecretsResponse{}, recipes.NewRecipeError(recipes.LoadSecretsFailed, fmt.Sprintf("failed to fetch secrets from the secret store resource id %s for Terraform recipe %s deployment: %s", secretStore, definition.TemplatePath, err.Error()), util.RecipeSetupError, recipes.GetErrorDetails(err))
+			// Get list of secretStoreIDs to log in the error message.
+			secretStoreIDKeys := []string{}
+			for secretStoreIDKey := range secretStoreIDResourceKeys {
+				secretStoreIDKeys = append(secretStoreIDKeys, secretStoreIDKey)
+			}
+
+			return nil, recipes.NewRecipeError(recipes.LoadSecretsFailed, fmt.Sprintf("failed to fetch secrets from the secret store resource id %s for Terraform recipe %s deployment: %s", strings.Join(secretStoreIDKeys, ", "), definition.TemplatePath, err.Error()), util.RecipeSetupError, recipes.GetErrorDetails(err))
```
Comment (vishwahiremat) [inline pkg/recipes/engine/engine.go:250]: I am trying to understadn this better, is this processing required here? i think if we want add more details to the error message we should do it in LoadSecrets?

Comment (vishwahiremat) [inline pkg/recipes/engine/engine.go:250]: unit tests for this one would be helpful, in the error message i see we are listing all the secret stores ids. It maybe the case that LoadSecrets failed for only one secret store

File: pkg/recipes/terraform/config/providers/types_test.go [hunk 10]
```diff
@@ -94,12 +95,202 @@ func TestGetRecipeProviderConfigs(t *testing.T) {
...
+								{
+									AdditionalProperties: map[string]any{
+										"subscriptionid": 1234,
+										"tenant_id":      "745fg88bf-86f1-41af-43ut",
+									},
+									Secrets: map[string]datamodel.SecretReference{
+										"secret1": {
+											Source: "secretstoreid1",
+											Key:    "secretkey1",
+										},
```
Comment (vishwahiremat) [inline pkg/recipes/terraform/config/providers/types_test.go:171]: can we also add a test convering: empty AdditionalProperties and non empty secrets

File: pkg/recipes/terraform/config/config.go [hunk 11]
```diff
@@ -297,3 +300,31 @@ func (cfg *TerraformConfig) AddOutputs(localModuleName string) error {
...
+
+// addSecretKeys is helper function to update the inputSecretIDs map
+func addSecretKeys(inputSecretIDs map[string][]string, secretSourceID, key string) {
+	if _, ok := inputSecretIDs[secretSourceID]; !ok {
+		inputSecretIDs[secretSourceID] = []string{key}
+	} else {
+		inputSecretIDs[secretSourceID] = append(inputSecretIDs[secretSourceID], key)
+	}
+}
```
[the remark this answers is not in the document]
Comment (lakshmimsft) [reply pkg/recipes/terraform/config/config.go:330]: the functions mainly traverse the config to retrieve Ids. This is why I placed them in the config package. Let me know if that is not okay.

File: pkg/recipes/driver/terraform.go [hunk 12]
```diff
@@ -74,6 +75,7 @@ type terraformDriver struct {
 // Execute creates a unique directory for each execution of terraform and deploys the recipe using the
 // the Terraform CLI through terraform-exec. It returns a RecipeOutput or an error if the deployment fails.
 func (d *terraformDriver) Execute(ctx context.Context, opts ExecuteOptions) (*recipes.RecipeOutput, error) {
+	var secrets map[string]string
```
[the remark this answers is not in the document]
Comment (lakshmimsft) [reply pkg/recipes/driver/terraform.go]: updated function with unsetGitConfigForDirIfApplicable() as discussed. pls take a look.

File: pkg/recipes/engine/engine.go (hunk 9 above)
Comment (lakshmimsft) [reply pkg/recipes/engine/engine.go]: added test here: https://github.com/radius-project/radius/pull/7695/files?diff=unified&w=0#diff-3fdb9931c1f2ddfc5b14d87f47b4124f2bdc1da187219fa1ad9cbe79fc326528L319

Comment (lakshmimsft) [reply pkg/recipes/engine/engine.go]: discussed with Vishwa, added test https://github.com/radius-project/radius/pull/7695/files?diff=unified&w=0#diff-3fdb9931c1f2ddfc5b14d87f47b4124f2bdc1da187219fa1ad9cbe79fc326528L319

File: pkg/recipes/driver/terraform_test.go [hunk 13]
```diff
@@ -781,3 +781,239 @@ func Test_Terraform_PrepareRecipeResponse(t *testing.T) {
...
+			envConfig:     createTerraformConfigWithEnvSecrets(),
+			definition:    definition,
+			expectedError: false,
+			expectedSecretIDs: map[string][]string{
+				"secret-store-id1":    {"secret-key-env2"},
+				"secret-store-id-env": {"secret-key-env1"},
+			},
+		},
+		{
+			name:          "GetSecretStoreID returns error",
```
Comment (vishwahiremat) [inline pkg/recipes/driver/terraform_test.go:841]:
```suggestion
			name:          "GetPrivateGitRepoSecretStoreID returns error",
```

**Review by vishwahiremat — DISMISSED** [review]
lgtm

File: pkg/recipes/configloader/secrets.go [hunk 14]
```diff
@@ -34,21 +35,41 @@ type secretsLoader struct {
...
 }
 
-func (e *secretsLoader) LoadSecrets(ctx context.Context, secretStore string) (v20231001preview.SecretStoresClientListSecretsResponse, error) {
-	secretStoreID, err := resources.ParseResource(secretStore)
-	if err != nil {
-		return v20231001preview.SecretStoresClientListSecretsResponse{}, err
-	}
+// LoadSecrets loads secrets from secret stores based on input map of provided secret store IDs and secret keys.
+// It returns a map of secret data, where the keys are the secret store IDs and the values are maps of secret keys and their corresponding values.
+func (e *secretsLoader) LoadSecrets(ctx context.Context, secretStoreIDResourceKeys map[string][]string) (secretData map[string]map[string]string, err error) {
```
[the remark this answers is not in the document]
Comment (lakshmimsft) [reply pkg/recipes/configloader/secrets.go:40]: pulled code after call to client.ListSecrets() into another function and created unit tests for populateSecretData() as discussed

File: pkg/recipes/terraform/config/config.go (hunk 11 above)
[the remark this answers is not in the document]
Comment (lakshmimsft) [reply pkg/recipes/terraform/config/config.go]: As discussed, moved this to terraform/types.go. I also refactored the functions. Pls take another look.

**Review by vishwahiremat — APPROVED** [review]

File: pkg/recipes/terraform/config/providers/types.go (hunk 1 above)
Comment (willtsai) [reply pkg/recipes/terraform/config/providers/types.go]: taking from a (hotly debated) example from the Hasicorp Vault project, seems like they opted to keep the overwriting behavior with the behavior documented:
- https://github.com/hashicorp/vault/issues/182
- https://github.com/hashicorp/vault/commit/2d9b12b85306290b4b58cd764cf0f8fa0d0f5b17

Thus, I think it's ok to proceed with the overwriting behavior.

Comment (lakshmimsft) [reply pkg/recipes/terraform/config/providers/types.go]: Thanks @willtsai. I have added a test for this case. We'll get @kachawla input on this and then I will document this behaviour in the design doc.

File: pkg/recipes/configloader/secrets.go [hunk 15]
```diff
@@ -34,21 +35,55 @@ type secretsLoader struct {
...
+	if secrets == nil {
+		return nil, fmt.Errorf("secrets not found for secret store ID '%s'", secretStoreID)
 	}
 
-	secrets, err := client.ListSecrets(ctx, secretStoreID.Name(), map[string]any{}, nil)
-	if err != nil {
-		return v20231001preview.SecretStoresClientListSecretsResponse{}, err
+	for _, secretKey := range secretKeys {
+		secretDataValue, ok := secrets.Data[secretKey]
+		if ok {
```
Comment (ytimocin) [inline pkg/recipes/configloader/secrets.go:78]: Why not inline this so that it can look clearer?

File: pkg/recipes/configloader/secrets_test.go [hunk 16]
```diff
@@ -0,0 +1,81 @@
...
+			} else {
+				require.NoError(t, err)
+				require.Equal(t, tt.expectedSecrets, secretData)
+			}
+		})
+	}
+}
+
+// Helper function to return a pointer to a string
+func ptrToString(s string) *string {
```
Comment (ytimocin) [inline pkg/recipes/configloader/secrets_test.go:79]: We should have this function somewhere else. If not, why not use the `to` library?

File: pkg/recipes/driver/gitconfig_test.go [hunk 17]
```diff
@@ -128,19 +126,12 @@ func TestUnsetGitConfigForDir(t *testing.T) {
...
-				"username": {
-					Value: to.Ptr("test-user"),
-				},
-				"pat": {
-					Value: to.Ptr("ghp_token"),
-				},
-			},
-		},
+func getSecretList() map[string]string {
+	secrets := map[string]string{
```
Comment (ytimocin) [inline pkg/recipes/driver/gitconfig_test.go:130]: Would we need to test multiple secrets?

Comment (lakshmimsft) [reply pkg/recipes/driver/gitconfig_test.go:130]: We will test for only single secret at this time in gitconfig_test.go. gitconfig.go is working with a single secret for private git repository. I'll defer to @vishwahiremat for more details.

File: pkg/recipes/configloader/secrets.go (hunk 15 above)
Comment (lakshmimsft) [reply pkg/recipes/configloader/secrets.go]: updated

**Review by ytimocin — APPROVED** [review]
